qadence_libs/qinfo_tools/qng.py

```diff
@@ -0,0 +1,181 @@
+        if 0.0 >= beta:
+            raise ValueError(f"Invalid beta value: {beta}")
+        if 0.0 > epsilon:
+            raise ValueError(f"Invalid epsilon value: {epsilon}")
+
+        if isinstance(model, TransformedModule):
+            logger.warning(
+                "The model is of type '<class TransformedModule>. "
+                "Keep in mind that the QNG optimizer can only optimize circuit "
+                "parameters. Input and output shifting/scaling parameters will not be optimized."
+            )
+            # Retrieve the quantum model from the TransformedModule
+            model = model.model
+        if not isinstance(model, QuantumModel):
+            raise TypeError(
+                "The model should be an instance of '<class QuantumModel>' "
+                f"or '<class TransformedModule>'. Got {type(model)}."
+            )
+
+        self.param_dict = model.vparams
+        self.circuit = model._circuit.abstract
+        if not isinstance(self.circuit, QuantumCircuit):
+            raise TypeError(
+                "The circuit should be an instance of '<class QuantumCircuit>'."
+                "Got {type(self.circuit)}"
+            )
+
+        defaults = dict(
+            model=model,
+            lr=lr,
+            approximation=approximation,
+            beta=beta,
+            epsilon=epsilon,
+        )
+        super(QuantumNaturalGradient, self).__init__(params, defaults)
+
+        if approximation == FisherApproximation.SPSA:
+            state = self.state["state"]
+            state.setdefault("iter", 0)
+            state.setdefault("qfi_estimator", None)
+
+    def step(self, closure: Callable | None = None) -> torch.Tensor:
+        """Performs a single optimization step of the QNG algorithm.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            vparams_values = [p for p in group["params"] if p.requires_grad]
+
+            # Build the parameter dictionary
+            # We rely on the `vparam()` method in `QuantumModel` and the
+            # `parameters()` in `nn.Module` to give the same param ordering.
+            # We test for this in `test_qng.py`.
+            vparams_dict = dict(zip(self.param_dict.keys(), vparams_values))
+
+            approximation = group["approximation"]
+            grad_vec = torch.tensor([v.grad.data for v in vparams_values])
+            if approximation == FisherApproximation.EXACT:
+                # Calculate the EXACT metric tensor
+                metric_tensor = 0.25 * get_quantum_fisher(
+                    self.circuit,
+                    vparams_dict=vparams_dict,
+                )
+
+                with torch.no_grad():
+                    # Finite shift the metric tensor to avoid problems when inverting
```

Comment:
What problems ?

Reply:
Numerical stability issues when inverting the matrix. We are no longer inverting the metric tensor (we solve the least square problems instead) but the shift is still needed for stability. I will change the comment accordingly.